falcon_mutualauth/mutualauth.py

```diff
@@ -74,11 +74,35 @@ def __init__(self, *default_roles):
     # NOTE(fxfitz): Falcon v1.0 requires the params parameter, so adding this
     # here now to be forward compatible with Falcon when v1.0 is released
     def process_resource(self, req, resp, resource, params=None):
-        authorized_roles = getattr(
-            resource, 'authorized_for', self._default_roles)
+        authd_roles = getattr(resource, 'authorized_for_map', {})
+
         roles = set(req.get_header('X-User-Roles').split(','))
 
-        if roles.isdisjoint(authorized_roles):
+        get_roles = authd_roles.get('on_get', self._default_roles)
+        post_roles = authd_roles.get('on_post', self._default_roles)
+        put_roles = authd_roles.get('on_put', self._default_roles)
+        del_roles = authd_roles.get('on_del', self._default_roles)
+
+        no_get_auth = roles.isdisjoint(get_roles)
+        no_put_auth = roles.isdisjoint(post_roles)
+        no_del_auth = roles.isdisjoint(put_roles)
+        no_post_auth = roles.isdisjoint(del_roles)
+
+        is_authorized = True
+
+        if no_get_auth and req.method == 'GET':
+            is_authorized = False
+
+        if no_put_auth and req.method == 'PUT':
+            is_authorized = False
+
+        if no_del_auth and req.method == 'DELETE':
+            is_authorized = False
+
+        if no_post_auth and req.method == 'POST':
+            is_authorized = False
+
+        if not is_authorized:
```

Comment:
I don't understand why this function computes all of the roles and whether the user has them. Semantically, you're attaching roles to (method, resource) tuples instead of just a resource. This class implements the resource part, so I'd expect this method to do the method dispatch. Specifically, that would mean:

* get the key for the appropriate roles for the current method (`"on_{method}".format(method=req.method.lower())`)
* `not_authorized = roles.is_disjoint(needed_roles)` or whatever

That way, this method doesn't actually know about which HTTP methods exist (useful, because e.g. PATCH), and will be much terser.